taf/git.py

```diff
@@ -33,34 +34,65 @@ def __init__(self, repo_path, repo_urls=None, additional_info=None):
   def is_git_repository(self):
     return (Path(self.repo_path) / '.git').is_dir()
 
-
-  def _git(self, cmd, *args):
+  def _git(self, cmd, *args, **kwargs):
     """Call git commands in subprocess
-
-    E.G.:
+    e.g.:
       self._git('checkout {}', branch_name)
     """
+    log_error = kwargs.pop('log_error', False)
+    log_error_msg = kwargs.pop('log_error_msg', '')
+    reraise_error = kwargs.pop('reraise_error', False)
+    log_success_msg = kwargs.pop('log_success_msg', '')
+
     if len(args):
       cmd = cmd.format(*args)
-    return run('git -C {} {}'.format(self.repo_path, cmd))
+    command = 'git -C {} {}'.format(self.repo_path, cmd)
+
+    if log_error or log_error_msg:
```

Comment:
I think passing _logging_ arguments to `_git` method is little bit confusing.

I think you can implement something similar with decorator and use it instead:

```python
def log(success_msg, error_msg, re_raise=False):
    def _log(f):
        def wrapper(self, *args):
            try:
                logger.debug("Repo %s: %s", self.repo_name, success_msg)
                res = f(self, *args)
                logger.debug("After repo %s: %s", self.repo_name, res)
                return res
            except:
                if error_msg:
                    logger.error("Repo %s: %s", self.repo_name, error_msg)
                if re_raise:
                    raise
        return wrapper
    return _log
```

`run` function should always log calls and we will see something like:
`git -c path/to/repo commit -m "Message"`, so maybe that's just enough for now.

Reply:
And use it like this:

```python
@log(
    'Check out branch',
    'Failed to check out branch',
    re_raise=True
)
def checkout_branch(self, branch_name, create=False):
    """..."""
```

Log output e.g.: "Repo taf: Check out branch ['master']".

Reply:
There seem to be some libraries for this. See https://www.sighalt.de/remove-visual-noise-of-logging-code-by-using-python-decorators.html

Reply:
I was looking at [this method](https://github.com/openlawlibrary/taf/blob/11776f22f1cdcb2c7db9f4190b646ee827e51d7c/taf/git.py#L70) for example. Looking just at the log file, we won't know if this method is called or some other, because `_git` method will log something like `Repo my-repo: rev-list 123abc..HEAD`, but I might be wrong. Anyway, since it's not priority and it does the job I am fine with that.

Reply:
With the current setup, we can add calls to methods which call `_git`. Let's add an issue for improving our logging strategy and address it later, when there aren't as many high-priority tasks.

Reply:
Done: https://github.com/openlawlibrary/taf/issues/27